patches/api/0332-Add-Recipe-API.patch

```diff
@@ -0,0 +1,235 @@
++    /**
++     * Recipes for smelting inside of a {@link org.bukkit.Material#FURNACE FURNACE}.
++     */
++    public static final RecipeType<FurnaceRecipe, Inventory> SMELTING = register("smelting");
++    /**
++     * Recipes for smelting inside of a {@link org.bukkit.Material#BLAST_FURNACE BLAST_FURNACE}.
++     */
++    public static final RecipeType<BlastingRecipe, Inventory> BLASTING = register("blasting");
++    /**
++     * Recipes for smelting inside of a {@link org.bukkit.Material#SMOKER SMOKER}.
++     */
++    public static final RecipeType<SmokingRecipe, Inventory> SMOKING = register("smoking");
++    /**
++     * Recipes for smelting on a {@link org.bukkit.Material#CAMPFIRE CAMPFIRE}.
++     */
++    public static final RecipeType<CampfireRecipe, Inventory> CAMPFIRE_COOKING = register("campfire_cooking");
++    /**
++     * Recipes for cutting in a {@link org.bukkit.Material#STONECUTTER STONECUTTER}.
++     */
++    public static final RecipeType<StonecuttingRecipe, Inventory> STONECUTTING = register("stonecutting");
++    /**
++     * Recipes for upgrading items in a {@link org.bukkit.Material#SMITHING_TABLE SMITHING_TABLE}.
++     */
++    public static final RecipeType<SmithingRecipe, Inventory> SMITHING = register("smithing");
++
++    /**
++     * Gets a registered recipe type by its key.
++     * @return a registered recipe type or null
++     */
++    public static @Nullable RecipeType<?, ?> getByKey(final @NotNull NamespacedKey key) {
++        return recipeTypes.get(Preconditions.checkNotNull(key, "key"));
++    }
++
++    /**
++     * Gets an iterator across all registered recipe types.
++     * @return recipe type iterator
++     */
++    public static @NotNull Iterator<RecipeType<?, ?>> iterator() {
++        return recipeTypes.values().iterator();
++    }
++}
+diff --git a/src/main/java/org/bukkit/Registry.java b/src/main/java/org/bukkit/Registry.java
+index a696fcaffa03af9e6c92e2ef3e12b38eb59e5db4..fb082e495fa1e9c3392f88431c0a37b0bbf33c2c 100644
+--- a/src/main/java/org/bukkit/Registry.java
++++ b/src/main/java/org/bukkit/Registry.java
+@@ -189,6 +189,26 @@ public interface Registry<T extends Keyed> extends Iterable<T> {
+             return GameEvent.getByKey(key);
+         }
+     };
++    // Paper start
++    /**
++     * Recipe types.
++     *
++     * @see io.papermc.paper.inventory.RecipeType RecipeType
++     */
++    Registry<io.papermc.paper.inventory.RecipeType<?, ?>> RECIPE_TYPE = new Registry<>() {
++        @NotNull
++        @Override
++        public Iterator<io.papermc.paper.inventory.RecipeType<?, ?>> iterator() {
++            return io.papermc.paper.inventory.RecipeType.iterator();
++        }
++
++        @Nullable
++        @Override
++        public io.papermc.paper.inventory.RecipeType<?, ?> get(@org.jetbrains.annotations.NotNull NamespacedKey key) {
++            return io.papermc.paper.inventory.RecipeType.getByKey(key);
++        }
++    };
++    // Paper end
```

Comment:
I hesitate to add new registries to upstream's Registry interface. I recently moved a couple new additions in PRs of mine to a PaperRegistry that just extended this registry. My thinking is that if upstream ever deigns to add something here, and paper now has a duplicate, the Registry class won't get cluttered up with deprecated static fields

Reply:
i kinda doubt that'll happen here, given that upstream went for the lazy approach of only supporting 1 type of crafting inventory, but i'd be happy to use the PaperRegistry class if/when it's merged :p